### PR TITLE
update to modern docker package

### PR DESCRIPTION
**Pull request by liudonghua123**

### CODE REVIEW

File: entrypoint.py [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
 
 class ImageNotFound(Exception):
     pass
 
-class MainObj(object):
+
+class Main:
     def __init__(self):
-        super(MainObj, self).__init__()
+        super(Main, self).__init__()
```
Comment (lukapeschke) [inline entrypoint.py]:
```suggestion
        super().__init__()
```

Heh, Python has changed a lot since I wrote this :)

File: entrypoint.py [hunk 2]
```diff
@@ -22,33 +27,35 @@ def _print_cmds(self):
...
 
     def _insert_step(self, step):
+        # ignore the end "# buildkit" comment
+        step = re.sub("\s*#.+$", "", step)
         if "#(nop)" in step:
             to_add = step.split("#(nop) ")[1]
         else:
-            to_add = ("RUN {}".format(step))
+            # step may contains "/bin/sh -c ", just ignore it
+            to_add = "{}".format(step.replace("/bin/sh -c ", ""))
```
Comment (lukapeschke) [inline entrypoint.py]: Maybe this should be changed in another PR ? The results with the current docker image and one built from your branch are quite different:

With the current image:

```
❯ docker run --rm -v '/var/run/docker.sock:/var/run/docker.sock' lukapeschke/dfa 21ab84251c7b
FROM lukapeschke/dfa:latest
ADD file:093f0723fa46f6cdbd6f7bd146448bb70ecce54254c35701feeceb956414622f in /
CMD ["/bin/sh"]
RUN /bin/sh -c apk add --update python3 wget      \
    && wget -O - --no-check-certificate https://bootstrap.pypa.io/get-pip.py | python3      \
    && apk del wget      \
    && pip3 install -U docker-py      \
    && yes | pip3 uninstall pip
COPY file:df2a17029ab5f655279e92b51f29046f182eb0a4a5457ad06b2338387bcbbd20 in /root
ENTRYPOINT ["/root/entrypoint.py"]
```

With the image build from your PR:

```
❯ docker run --rm -v '/var/run/docker.sock:/var/run/docker.sock' updated-image 21ab84251c7b
FROM lukapeschke/dfa:latest
/bin/sh -c
/bin/sh -c
apk add --update python3 wget      \
    && wget -O - --no-check-certificate https://bootstrap.pypa.io/get-pip.py | python3      \
    && apk del wget      \
    && pip3 install -U docker-py      \
    && yes | pip3 uninstall pip
/bin/sh -c
/bin/sh -c
```

**Review by lukapeschke — CHANGES_REQUESTED** [review]
Thanks a lot for creating this! The changes to how the image is retrieved and to the client lib usage look good to me, but the changes to `_insert_step` seem to break the existing behaviour... Maybe those should be done in another PR ?

File: entrypoint.py (hunk 2 above)
Comment (liudonghua123) [reply entrypoint.py]: I updated the code 😄 It should fixed in the following commits.

File: entrypoint.py (hunk 1 above)
Comment (liudonghua123) [reply entrypoint.py]: Yes, I didn't notice this change.

Comment (liudonghua123) [reply entrypoint.py]: Because `Main` is not extend any explicit base class, so `super().__init__()` is optional and recommend to ignore it IMO.

File: entrypoint.py [hunk 3]
```diff
@@ -22,33 +26,35 @@ def _print_cmds(self):
...
     def _insert_step(self, step):
+        # ignore the end "# buildkit" comment
+        step = re.sub("\s*# buildkit$", "", step)
         if "#(nop)" in step:
-            to_add = step.split("#(nop) ")[1]
+            to_add = step.split("#(nop)")[1].strip()
         else:
-            to_add = ("RUN {}".format(step))
+            # step may contains "/bin/sh -c ", just ignore it
+            to_add = "{}".format(step.replace("/bin/sh -c ", ""))
```
Comment (lukapeschke) [inline entrypoint.py]:
```suggestion
            to_add = "RUN {}".format(step.replace("/bin/sh -c ", ""))
```

**Review by lukapeschke — APPROVED** [review]
Just fixed a missing `RUN`, script output looks good now :slightly_smiling_face: Thanks a lot for your contribution!

File: entrypoint.py (hunk 3 above)
Comment (liudonghua123) [reply entrypoint.py]: If add `RUN ` here, some docker commands result redundant `RUN ` output.

image

Comment (lukapeschke) [reply entrypoint.py]: Hmm that's really weird :thinking: , I can't reproduce it on my side... I've noted myself to add some proper versioning + CI when I have the time: https://github.com/lukapeschke/dockerfile-from-image/issues/4
